pkg/rancher/migrations.go

```diff
@@ -736,6 +746,51 @@ func migrateHarvesterCloudCredentialExpiration(w *wrangler.Context) error {
 	return createOrUpdateConfigMap(w.Core.ConfigMap(), cm)
 }
 
+// migrateImportedClusterFields will update fields on clusters.management.cattle.io/v3 for imported clusters,
+// fields will no longer be set by clusters.provisioning.cattle.io/v1 ensuring management cluster
+// remains the source of truth for imported clusters.
+func migrateImportedClusterFields(w *wrangler.Context) error {
+	cm, err := getConfigMap(w.Core.ConfigMap(), migrateImportedClusterManagedFields)
+	if err != nil || cm == nil {
+		return err
+	}
+
+	if cm.Data[importedClusterManagedFieldsMigratedKey] == "true" {
+		return nil
+	}
+
+	provClusters, err := w.Provisioning.Cluster().List("", metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, cluster := range provClusters.Items {
+		if mgmtNameRegexp.MatchString(cluster.Name) || cluster.Spec.RKEConfig != nil {
+			// run only for non-legacy imported clusters
+			continue
+		}
+		mgmtCluster, err := w.Mgmt.Cluster().Get(cluster.Status.ClusterName, metav1.GetOptions{})
+		if k8serror.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			return err
+		}
+		clusterCopy := mgmtCluster.DeepCopy()
+		// clusterDeploy controller uses GetDesiredAgentImage(mgmtCluster), customizable using Spec.DesiredAgentImage
+		clusterCopy.Spec.DesiredAgentImage = ""
+		// clusterDeploy controller uses GetDesiredAuthImage(mgmtCluster), customizable using Spec.DesiredAuthImage
+		clusterCopy.Spec.DesiredAuthImage = ""
+		// managed by Spec.ImportedConfig.PrivateRegistryURL for imported clusters
```

Comment:
it looks like `clusterDeploy` sets those two fields only if their value is "fixed" ( [code](https://github.com/rancher/rancher/blob/main/pkg/controllers/management/clusterdeploy/clusterdeploy.go#L499-L505)),  so I wonder whether we need to set them "fixed" so that they will be rest to the expected new values? Or is there another logic to correct those two fields that I am missing?

Reply:
Yep, imo that's expected behavior. DesiredAgent(/Auth)Images are only meant to be set by user for management clusters, that is why we update them on status only when configured. 

But these fields being empty is the default behavior, which fetches agent image from settings: https://github.com/rancher/rancher/blob/3aaf47cecc730f4ff22030c309a5ec328036af9e/pkg/systemtemplate/import.go#L308 and sets on status as expected: https://github.com/rancher/rancher/blob/3aaf47cecc730f4ff22030c309a5ec328036af9e/pkg/controllers/management/clusterdeploy/clusterdeploy.go#L497-L498